### ISSUE TITLE
Plans to add Gemini?

### ISSUE THREAD

**Issue by satonotdead** [merefield/discourse-chatbot#20]
Bard can search the web for free! 🤟

![image](https://github.com/merefield/discourse-chatbot/assets/98671738/44837872-8baf-4fe6-b520-df27b6e75dcd)

![image](https://github.com/merefield/discourse-chatbot/assets/98671738/62929a3e-e0fc-4642-bbbe-9c94adf09e1b)

https://github.com/ra83205/google-bard-api

**Comment by merefield** [comment]
Yes, I'd love to add support and for any emerging primary publicly-accessible LLM from Google.

Not whether but when I get around to it.

PR welcome in meantime!

**Comment by merefield** [comment]
Apologies, on review, I'm removing this from roadmap as given the general lack of community PRs I do not have capacity to entertain it.

**Comment by dillfrescott** [comment]
Awwww I think this would have been awesome!

**Comment by merefield** [comment]
I would accept a PR @dillfrescott if it's not a settings nightmare. If someone is willing to fund me I would consider doing it myself.  Renaming the issue.

**Comment by dillfrescott** [comment]
Hey Robert thanks for the response. Is there a rough estimate of how much someone would have to pay for you to integrate it?

**Comment by dillfrescott** [comment]
Found [this](https://github.com/PublicAffairs/openai-gemini). Do you think that could work instead of having to integrate it into the plugin?

**Comment by dillfrescott** [comment]
Update: It works fantastically! updating the `chatbot.prompt.post` value to just `%{raw}` works great for Gemini!

**Comment by merefield** [comment]
That's nuts @dillfrescott!! ❤️. Great find!

And there's me thinking that proxies were just there to add Auth headers 😂

I wonder if there's a Claude equivalent?! : 🤔

**Comment by merefield** [comment]
> Update: It works fantastically! updating the `chatbot.prompt.post` value to just `%{raw}` works great for Gemini!

Does Gemini struggle with usernames being included?  Can you elaborate on what effect that has?  Doesn't excluding them take away the ability of the bot to @mention?

**Comment by dillfrescott** [comment]
Hmm indeed was having issues with mentions. I hadn't tested that.

I changed the prompt to:

```
%{username} said:

%{raw}

You are a helpful assistant. Please respond to what %{username} said appropriately.
```

And now it works much better!

![Screenshot 2024-03-09 035753](https://github.com/merefield/discourse-chatbot/assets/56714680/2f2bf9de-29f6-493c-b68e-aabffb3da645)

**Comment by dillfrescott** [comment]
> I wonder if there's a Claude equivalent?

Honestly probably 😅

**Comment by dillfrescott** [comment]
![aaaaaa](https://github.com/merefield/discourse-chatbot/assets/56714680/eb91f481-c428-400b-bbbf-57afe23bd72e)

As you can see it also handles topic mention requests quite well!

**Comment by merefield** [comment]
Thinking about this some more, @dillfrescott, how is this arrangement going to handle Embeddings?  That's critical to getting the full power out of Discourse Chatbot.

**Comment by dillfrescott** [comment]
I don't think it can do embeddings yet, at least with this setup. I am using a different proxy though than the one I linked, currently.

**Comment by dillfrescott** [comment]
I would totally use openai's API but gemini API is totally free to use and openai's is not.

**Comment by dillfrescott** [comment]
[This project](https://github.com/mth93/langchain-gemini-pro-proxy) seems to be able to also do embeddings already. It seems to be only one python file too, so easy to run.

**Comment by dillfrescott** [comment]
I think [this project](https://github.com/dillfrescott/langchain-gemini-pro-proxy) can do embeddings from a quick glance at it.

**Comment by merefield** [comment]
> I would totally use openai's API but gemini API is totally free to use and openai's is not.

I understand, though remember to manage your spend you can use the inbuilt quota system.

**Comment by merefield** [comment]
> I think [this project](https://github.com/dillfrescott/langchain-gemini-pro-proxy) can do embeddings from a quick glance at it.

that's a 404 for me

**Comment by dillfrescott** [comment]
Wow the author deleted it!

**Comment by dillfrescott** [comment]
my bad i linked a fork i deleted after testing it!

https://github.com/mth93/langchain-gemini-pro-proxy

**Comment by merefield** [comment]
I really like your original suggestion though as it has been package nicely with cloud deployment buttons.  Let me know if you PR embeddings and function support ;)

**Comment by merefield** [comment]
I've added the original example and the approach to the Readme

**Comment by merefield** [comment]
For basic chat interaction, use base URL (aka `Chatbot open AI model custom URL high trust`) for compatible LLM services that support the OpenAI API standard.  Closing this issue.
